courses/models.py

```diff
@@ -0,0 +1,799 @@
+        Checks if the course run in the past
+
+        Returns:
+            boolean: True if course run has ended
+
+        """
+        if not self.end_date:
+            return False
+        return self.end_date < now_in_utc()
+
+    @property
+    def is_not_beyond_enrollment(self):
+        """
+        Checks if the course is not beyond its enrollment period
+
+
+        Returns:
+            boolean: True if enrollment period has begun but not ended
+        """
+        now = now_in_utc()
+        return (
+            (self.end_date is None or self.end_date > now)
+            and (self.enrollment_end is None or self.enrollment_end > now)
+            and (self.enrollment_start is None or self.enrollment_start <= now)
+        )
+
+    @property
+    def is_unexpired(self):
+        """
+        Checks if the course is not expired
+
+        Returns:
+            boolean: True if course is not expired
+        """
+        return not self.is_past and self.is_not_beyond_enrollment
+
+    @property
+    def courseware_url(self):
+        """
+        Full URL for this CourseRun as it exists in the courseware
+
+        Returns:
+            str or None: Full URL or None
+        """
+        return (
+            edx_redirect_url(self.courseware_url_path)
+            if self.courseware_url_path
+            else None
+        )
+
+    @property
+    def text_id(self):
+        """Gets the courseware_id"""
+        return self.courseware_id
+
+    def __str__(self):
+        title = f"{self.courseware_id} | {self.title}"
+        return title if len(title) <= 100 else title[:97] + "..."
+
+    def clean(self):
+        """
+        If expiration_date is not set:
+        1. If end_date is provided: set expiration_date to default end_date + 90 days.
+        2. If end_date is None, don't do anything.
+
+        Validate that the expiration date is:
+        1. Later than end_date if end_date is set
+        2. Later than start_date if start_date is set
+        """
+        if not self.expiration_date:
+            return
+
+        if self.start_date and self.expiration_date < self.start_date:
+            raise ValidationError("Expiration date must be later than start date.")
+
+        if self.end_date and self.expiration_date < self.end_date:
+            raise ValidationError("Expiration date must be later than end date.")
+
+    def save(
+        self, force_insert=False, force_update=False, using=None, update_fields=None
+    ):
+        """
+        Overriding the save method to inject clean into it
+        """
+        self.clean()
+        super().save(
+            force_insert=force_insert,
+            force_update=force_update,
+            using=using,
+            update_fields=update_fields,
+        )
+
+
+class EnrollmentModel(TimestampedModel, AuditableModel):
+    """Abstract base model for enrollments"""
+
+    class Meta:
+        abstract = True
+
+    user = models.ForeignKey(User, on_delete=models.PROTECT)
```

Comment:
Here's one we could discuss: I think we should get rid of all `on_delete=models.PROTECT` rules from these models. The obvious drawback is that it's tremendously disruptive to the development and testing experience (at least for me it has been). The benefit is that we get some protection from errant deletions, but I don't know if that's something we really need in production given the way that this app and others like it are administered at OL. I think deletions in prod happen very rarely. If it's done through Django admin, which in practice it basically always is, you already get a confirmation screen that tells you what will be deleted before you pull the trigger.

Reply:
I went ahead and switched that to cascade, it's not like it's a DB-level protection anyway.